src/box/lua/net_box.lua

```diff
@@ -985,6 +987,70 @@ local console_mt = {
     __metatable = false
 }
 
+local index_wrap_mt = {
+    __index = function(self, key)
+        return self._src[key]
+    end
+}
+
+local function wrap_index(space, src)
+    return setmetatable({
+        _stream_id = space._stream_id,
+        _src = src,
+        _space = space,
+    }, index_wrap_mt)
+end
+
+local proxy_index_mt = {
+    __index = function(self, key)
+        local _space = self._space
+        local src = _space._src.index[key]
+        if not src then
+            return nil
+        end
+        local res = wrap_index(_space, src)
+        self[key] = res
+        return res
+    end
+}
+
+local space_wrap_mt = {
+    __index = function(self, key)
+        return self._src[key]
+    end
+}
+
+local function wrap_space(proxy, src)
+    local res = setmetatable({
+        _stream_id = proxy._stream_id,
+        _src = src,
+        index = setmetatable({
+            _space = nil,
+        }, proxy_index_mt)
+    }, space_wrap_mt)
+    res.index._space = res
+    return res
+end
+
+local proxy_space_mt
+proxy_space_mt = {
+    __index = function(self, key)
+        local proxy = self._proxy
+        local src = proxy._src.space[key]
+        if not src then
+            return nil
+        end
+        proxy._src:on_schema_reload(function()
+            proxy.space = setmetatable({
```

Comment:
It looks bad that we need a closure for each space. Please, make it installed only once for the entire stream. It won't do any good to reset the same `_proxy` multiple times on each schema reload.

Also with the closures the problem is that AFAIR their creation can't be jitted. But I don't remember. And can't propose anything better and simpler now except schema version comparison and lazy reload. Can be done later then. (The part above about setting the trigger only once needs to be done now.)

Reply:
Fixed